components/ScriptLibrary/include/LHVMTypes.h

```diff
@@ -0,0 +1,371 @@
+/******************************************************************************
+ * Copyright (c) 2018-2024 openblack developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *******************************************************************************/
+
+#pragma once
+
+#include <cassert>
+#include <cstdint>
+
+#include <array>
+#include <functional>
+#include <map>
+#include <string>
+
+namespace openblack::lhvm
+{
+
+static const std::array<std::string, 31> k_OpcodeNames = {
+    "END",    "JZ",   "PUSH", "POP",       "ADD",       "SYS",        "SUB",       "NEG",  "MUL", "DIV", "MOD",
+    "NOT",    "AND",  "OR",   "EQ",        "NE",        "GE",         "LE",        "GT",   "LT",  "JMP", "SLEEP",
+    "EXCEPT", "CAST", "RUN",  "ENDEXCEPT", "RETEXCEPT", "FAILEXCEPT", "BRKEXCEPT", "SWAP", "LINE"};
+
+enum class LHVMVersion : uint32_t
+{
+	BlackAndWhite = 7,
+	CreatureIsle = 8,
+	BlackAndWhite2 = 12,
+
+	Last = 0xFFFFFFFF
+};
+
+enum class DataType : uint32_t
+{
+	None = 0,
+	Int,
+	Float,
+	Vector,
+	Object,
+	Unk5,
+	Boolean,
+	Unk7,
+
+	Last = 0xFFFFFFFF
+};
+static_assert(sizeof(DataType) == 4);
+
+static const std::array<std::string, 8> k_DataTypeNames = {"NONE",   "INT",  "FLOAT",   "VECTOR",
+                                                           "OBJECT", "UNK2", "BOOLEAN", "UNK4"};
+
+static const std::array<std::string, 8> k_DataTypeChars = {"", "I", "F", "V", "O", "", "B", ""};
+
+class VMValue
+{
+public:
+	VMValue()
+	    : floatVal(0.0f)
+	{
+	}
+
+	explicit VMValue(float value)
+	    : floatVal(value)
+	{
+	}
+
+	explicit VMValue(int32_t value)
+	    : intVal(value)
+	{
+	}
+
+	explicit VMValue(uint32_t value)
+	    : uintVal(value)
+	{
+	}
+
+	union
+	{
+		float floatVal;
+		int32_t intVal;
+		uint32_t uintVal;
+	};
+};
+static_assert(sizeof(VMValue) == 4);
```

Comment:
Yes, however having it as a class is convenient because we can have specific constructors for the data type we want to store. Without these constructors, initializing the union would be much verbose.

Reply:
I see, you're doing a [Union-like class](https://en.cppreference.com/w/cpp/language/union#Union-like_classes). I',m curious if you've considered [std::variant](https://en.cppreference.com/w/cpp/utility/variant)

Reply:
Yes, I've considered it, but it cannot work because I need `VMValue` to be exactly 32 bits, and `variant` is always larger because it holds both value and actual type. (see here: https://stackoverflow.com/questions/45575892/why-is-sizeofstdvariant-the-same-size-as-a-struct-with-the-same-members).
Personally I don't like having a class to hold just a value, but the benefit of being able to inline the initialization is priceless. However, I could replace this class with a `struct` and use the struct initialization syntax; whould this be better in your opinion?

Reply:
I've replaced the class with a typedef union, surprisingly it worked seamless without changing any reference to the previous constructors.

Reply:
`struct` with custom initializers worked on all platforms.